### PR TITLE
Implement opcode SELFBALANCE

### LINKED ISSUE #53: Opcode: SELFBALANCE

### PR DESCRIPTION
**Pull request by JulianVentura**
Closes #53 
This PR implements the opcode SELFBALANCE with tests.

### CODE REVIEW

**Comment by FrancoGiachetta** [comment]
LGTM

**Review by FrancoGiachetta — APPROVED** [review]
LGTM

File: src/db.rs [hunk 1]
```diff
@@ -37,17 +37,39 @@ impl Db {
...
+        contracts.insert(hash, bytecode);
+
+        Self {
+            accounts,
+            contracts,
+            block_hashes,
+        }
+    }
+
+    pub fn with_accounts(self, new_accounts: HashMap<Address, DbAccount>) -> Self {
```
Comment (MegaRedHand) [inline src/db.rs:62]: Can we change this to receive a single account? Then we can just chain it when multiple accounts are needed.

File: src/db.rs [hunk 2]
```diff
@@ -37,17 +37,39 @@ impl Db {
...
             bytecode_hash: hash,
             ..Default::default()
         };
-        db.accounts.insert(address, account);
-        db.contracts.insert(hash, bytecode);
-        db
+
+        let mut contracts = self.contracts.clone();
+        let mut accounts = self.accounts.clone();
+        let block_hashes = self.block_hashes;
```
Comment (MegaRedHand) [inline src/db.rs:50]: No need to clone here. You can just declare `mut self`, modify that, and then return it.

File: src/db.rs [hunk 3]
```diff
@@ -37,17 +37,39 @@ impl Db {
...
+            accounts,
+            contracts,
+            block_hashes,
+        }
+    }
+
+    pub fn with_accounts(self, new_accounts: HashMap<Address, DbAccount>) -> Self {
+        let contracts = self.contracts;
+        let block_hashes = self.block_hashes;
+        let mut accounts = self.accounts.clone();
```
Comment (MegaRedHand) [inline src/db.rs:65]: Same here. Just declare `mut self`:

```rust
fn with_accounts(mut self, ...) -> Self {
    self.accounts.extend(...);
    self
}
```

File: src/syscall.rs [hunk 4]
```diff
@@ -193,6 +193,20 @@ impl<'c> SyscallContext<'c> {
...
+        match self.db.basic(address).unwrap() {
+            Some(a) => {
+                balance.hi = (a.balance >> 128).low_u128();
+                balance.lo = a.balance.low_u128();
+            }
+            None => {
+                balance.hi = 0;
+                balance.lo = 0;
+            }
+        };
```
Comment (MegaRedHand) [inline src/syscall.rs:207]: You can use `unwrap_or_default` to simplify this.

File: tests/evm.rs [hunk 5]
```diff
@@ -1244,3 +1219,61 @@ fn balance_static_gas_check() {
...
+
+#[test]
+fn selfbalance_gas_check() {
+    let operations = vec![Operation::SelfBalance];
+    let mut env = Env::default();
+    env.tx.gas_limit = 999_999;
+    let needed_gas = gas_cost::SELFBALANCE;
+
+    run_program_assert_gas_exact(operations, env, needed_gas as _);
+}
```
Comment (MegaRedHand) [inline tests/evm.rs:1279]: Can we add an example that checks that `SelfBalance` does the same as [`Address`, `Balance`] ?

**Review by MegaRedHand — CHANGES_REQUESTED** [review]

File: tests/evm.rs (hunk 5 above)
Comment (JulianVentura) [reply tests/evm.rs:1279]: Yes, that's a good one!

File: src/db.rs [hunk 6]
```diff
@@ -36,18 +36,23 @@ impl Db {
...
-        db
+
+        self.accounts.insert(address, account);
+        self.contracts.insert(hash, bytecode);
+
+        Self {
+            accounts: self.accounts,
+            contracts: self.contracts,
+            block_hashes: self.block_hashes,
+        }
```
Comment (MegaRedHand) [inline src/db.rs:55]:
```suggestion
        self
```

**Review by MegaRedHand — APPROVED** [review]
LGTM

**Pull request merged**